node/collation-generation/src/lib.rs

```diff
@@ -312,108 +344,29 @@ async fn handle_new_activations<Context>(
-						//
-						// As such, honest collators never produce an uncompressed PoV which starts with
-						// a compression magic number, which would lead validators to reject the collation.
-						if encoded_size > validation_data.max_pov_size as usize {
-							gum::debug!(
-								target: LOG_TARGET,
-								para_id = %scheduled_core.para_id,
-								size = encoded_size,
-								max_size = validation_data.max_pov_size,
-								"PoV exceeded maximum size"
-							);
-
-							return
-						}
-
-						pov
-					};
-
-					let pov_hash = pov.hash();
-
-					let signature_payload = collator_signature_payload(
-						&relay_parent,
-						&scheduled_core.para_id,
-						&persisted_validation_data_hash,
-						&pov_hash,
-						&validation_code_hash,
-					);
-
-					let erasure_root =
-						match erasure_root(n_validators, validation_data, pov.clone()) {
-							Ok(erasure_root) => erasure_root,
-							Err(err) => {
-								gum::error!(
-									target: LOG_TARGET,
-									para_id = %scheduled_core.para_id,
-									err = ?err,
-									"failed to calculate erasure root",
-								);
-								return
-							},
-						};
-
-					let commitments = CandidateCommitments {
-						upward_messages: collation.upward_messages,
-						horizontal_messages: collation.horizontal_messages,
-						new_validation_code: collation.new_validation_code,
-						head_data: collation.head_data,
-						processed_downward_messages: collation.processed_downward_messages,
-						hrmp_watermark: collation.hrmp_watermark,
-					};
-
-					let ccr = CandidateReceipt {
-						commitments_hash: commitments.hash(),
-						descriptor: CandidateDescriptor {
-							signature: task_config.key.sign(&signature_payload),
-							para_id: scheduled_core.para_id,
+					construct_and_distribute_receipt(
+						PreparedCollation {
+							collation,
+							para_id: task_config.para_id,
 							relay_parent,
-							collator: task_config.key.public(),
-							persisted_validation_data_hash,
-							pov_hash,
-							erasure_root,
-							para_head: commitments.head_data.hash(),
+							validation_data,
 							validation_code_hash,
+							n_validators,
 						},
-					};
-
-					gum::debug!(
-						target: LOG_TARGET,
-						candidate_hash = ?ccr.hash(),
-						?pov_hash,
-						?relay_parent,
-						para_id = %scheduled_core.para_id,
-						"candidate is generated",
-					);
-					metrics.on_collation_generated();
-
-					if let Err(err) = task_sender
-						.send(
-							CollatorProtocolMessage::DistributeCollation(
-								ccr,
-								parent_head_data_hash,
-								pov,
-								result_sender,
-							)
-							.into(),
-						)
-						.await
-					{
-						gum::warn!(
-							target: LOG_TARGET,
-							para_id = %scheduled_core.para_id,
-							err = ?err,
-							"failed to send collation result",
-						);
-					}
+						task_config.key.clone(),
+						|msg| async move {
+							if let Err(err) = task_sender.send(msg.into()).await {
```

Comment:
Is the `sender` parameter to `handle_new_activations` the same as `ctx.sender()`? 

If so, perhaps we could clone `task_sender` from `ctx.sender()` and get rid of the `sender` parameter entirely.

Reply:
Yeah, that is the obvious next step of refactoring, but I left it beyond the scope of the PR. This code was written before we introduced the `SubsystemSender` trait mechanics.

Reply:
I had some spare time so just did this refactoring.